core/store/src/db/rocksdb/instance_tracker.rs

```diff
@@ -0,0 +1,171 @@
+use std::sync::{Condvar, Mutex};
+
+use once_cell::sync::Lazy;
+use tracing::info;
+
+/// Describes number of RocksDB instances and sum of their max_open_files.
+///
+/// This is modified by [`InstanceTracker`] which uses RAII to automatically
+/// update the trackers when new RocksDB instances are created and dropped.
+#[derive(Default)]
+struct Inner {
+    /// Number of RocksDB instances open.
+    count: usize,
+
+    /// Sum of max_open_files configuration option of all the RocksDB instances.
+    ///
+    /// This is used when opening multiple instances to guarantee that process
+    /// limit is high enough to allow all of the instances to open the maximum
+    /// number of files allowed in their respective configurations.
+    ///
+    /// For example, if NOFILE limit is set to 1024 and we try to open two
+    /// RocksDB instances with max_open_files set to 512 the second open should
+    /// fail since the limit is too low to accommodate both databases and other
+    /// file descriptor node is opening.
+    max_open_files: u64,
+}
+
+/// Synchronisation wrapper for accessing [`Inner`] singleton.
+#[derive(Default)]
+struct State {
+    mutex: Mutex<Inner>,
```

Comment:
I think `mutex` is not a very good name for this field.  I would propose that we call it `inner`.